README.md
# VideoDirectorJS

VideoDirectJS is a javascript library that makes it possible to link HTML5 video events/timing to your own (custom) callbacks.

## Loading support

* Classic loading. Loading through script src, objects accessible through VideoDirectJS namespace.
* AMD/Require.JS

## Usage

### Initialisation


```html
<video id="video-player" width="320" height="240" controls>
	<source src="video.mp4" type="video/mp4">
	Your browser does not support the video tag.
</video>
```

```javascript
var videoDirector = new VideoDirectorJS.Director('video-player');
```

### Schedule event

```javascript
// callback will be executed when video plays
videoDirector.at('play', function(){
  console.log('Video plays');
});

// callback will be executed when video is paused
videoDirector.at('pause', function(){
  console.log('Video pauses');
});

// callback will be executed when video ended
videoDirector.at('ended', function(){
  console.log('Video ended');
});

// callback will be executed when video position is at 00:00:02
videoDirector.at('2s', function(){
  console.log('Executed at 00:00:02');
});

// callback will be executed when video position is at 00:00:15
videoDirector.at('00:00:15', function(){
  console.log('Executed at 00:00:15');
});

// callback will be executed when video position is at 00:01:00
videoDirector.at('1m', function(){
  console.log('Executed at 00:01:00');
});

// callback will be executed when video position is at 00:01:00
videoDirector.at('00:01:00', function(){
  console.log('Executed at 00:01:00');
});

```
## Methods

### Director

#### at

Schedule video event

```javascript
// schedule video event when video ended
director.at('ended', function(){
	// video ended
});
```

#### playVideo

plays video

```javascript
// plays video
director.playVideo();
```

#### pauseVideo

pause video

```javascript
// pause video
director.pauseVideo();
```

#### muteVideo

mutes video audo

```javascript
// mutes video audio
director.muteVideo();
```

#### setVolume

set video volume

```javascript
// set volume to 100%
director.setVolume(100);

// set volume to 50%
director.setVolume(50);

// set volume to 20%
director.setVolume(20);
```

#### getCurrentTime

get current time in video

```javascript
// return seeked position (seconds)
director.getCurrentTime();
```

## Events

Passed to the **at** method

### Timed events

| timed event       | description                                                                                       |
| ----------------- | ------------------------------------------------------------------------------------------------- |
| 1s					| Called when video is at position 0:0:1                                                            |
| 1m					| Called when video is at position 0:1:0                                                            |
| 0:1:15				| Called when video is at position 0:1:15                                                           |


### Non timed events

| even name         | description                                                                                       |
| ----------------- | ------------------------------------------------------------------------------------------------- |
| play					| Executed when the video starts playing or is no longer paused                                     |
| playing				| Executed when a video starts to play after having been paused or stopped for buffering            |
| pause				| Executed when a video pauses                                                                      |
| ended 				| Executed when a video ended                                                                       |
| canplay 			| Executed when the browser can start playing the movie                                             |
| volumechange 		| Executed when the audio volume of a video has changed                                             |
| seeking     		| Executed when the user starts moving/skipping to a new position in the video                      |
| seeked      		| Executed when the user is finished moving/skipping to a new position in the video                 |


## License

GNU LESSER GENERAL PUBLIC LICENSE Version 2.1 (LGPL v2.1)


CHANGELOG
v1.1.0

- refactored code
- fixed tests
- added events: canplay, playing, volumechange

v1.2.0

- added get current time
- add method to mute video
- add method to set volume
- added events: seeked, seeking


tests/tests.js
/**
 * Testing videodirector exceptions
 */
QUnit.test('Test VideoDirector exceptions', function(assert) {
  //
  // None element passed
  assert.throws(
    function(){
      new VideoDirectorJS.Director('none-element')
    }, 'No element found with this id',
    'None element passed'
  );

  //
  // Wrong element passed
  assert.throws(
    function(){
      new VideoDirectorJS.Director('test-element1');
    }, 'Element is not a <video> element',
    'Wrong element passed'
  );

  //
  // Video hasn't initialised
  assert.throws(
    function(){
      var director =  new  VideoDirectorJS.Director('video-player');
      director.hasVideoInitialised = false;
      director.setupEvents();
    }, 'Video hasn\'t initialised',
    'Video hasn\'t initialised'
  );
});

/**
 * Testing video actions list
 */
QUnit.test('Test VideoDirector VideoActions list', function(assert) {
  // init video director
  var director =  new  VideoDirectorJS.Director('video-player');

  // schedule new action
  director.at('play', function(){});
  director.at('pause', function(){});
  director.at('10s', function(){});
  director.at('10s', function(){});
  director.at('10m', function(){});
  director.at('play', function(){});
  director.at('play', function(){});

  // test length
  assert.equal(director.getVideoActions().length, 4, 'Get number of video actions (length)');
  assert.equal(director.getTimedVideoActions().length, 3, 'Get number of timed video actions (length)');

  // get items
  assert.equal(director.getVideoActions()[0].getAction(), 'play', 'Expected action (play)');
  assert.equal(director.getVideoActions()[1].getAction(), 'pause', 'Expected action (pause)');
  assert.equal(director.getVideoActions()[2].getAction(), 'play', 'Expected action (play)');
  assert.equal(director.getVideoActions()[3].getAction(), 'play', 'Expected action (play)');

  // get videoactions of certain type
  assert.equal(director.getVideoActionsOfType('play').length, 3, 'Get VideoActionOfType (play: 3)');

  // get videoactions of certain time
  assert.equal(director.getTimedVideoActionsOfTime(10).length, 2, 'Get getTimedVideoActionsOfTime (10s)');
});

/**
 * Testing VideoActions
 */
QUnit.test('Test VideoActions', function(assert) {

  // create new video action
  var videoActionHello = new  VideoDirectorJS.VideoAction('play', function(){
    return 'hello world'
  });

  //test get action
  assert.equal(videoActionHello.getAction(), 'play',  'Get action');

  // test callback
  assert.equal(videoActionHello.executeCallback(), 'hello world',  'Execute callback');
});

/**
 * Testing VideoActions Parse timing
 */
QUnit.test('Test VideoActions parse time', function(assert) {
  // create video action
  var videoActionHello = new  VideoDirectorJS.VideoAction('10s', function(){});

  // test {xx}s structure
  videoActionHello.parseTimingAction('0s');
  assert.equal(videoActionHello.getTiming(), 0, 'Parse timing (0s)');

  videoActionHello.parseTimingAction('10s');
  assert.equal(videoActionHello.getTiming(), 10, 'Parse timing (10s)');

  videoActionHello.parseTimingAction('100s');
  assert.equal(videoActionHello.getTiming(), 100, 'Parse timing (100s)');

  // test {xx}m structure
  videoActionHello.parseTimingAction('1m');
  assert.equal(videoActionHello.getTiming(), 60, 'Parse timing (1m)');

  videoActionHello.parseTimingAction('5m');
  assert.equal(videoActionHello.getTiming(), 300, 'Parse timing (5m)');

  // test {xx}m {xx}s structure
  videoActionHello.parseTimingAction('1m30s');
  assert.equal(videoActionHello.getTiming(), 90, 'Parse timing (1m30s)');

  // test {xx}:{xx}:{xx} structure
  videoActionHello.parseTimingAction('0:0:30');
  assert.equal(videoActionHello.getTiming(), 30, 'Parse timing (0:0:30)');

  videoActionHello.parseTimingAction('0:0:4250');
  assert.equal(videoActionHello.getTiming(), 4250, 'Parse timing (0:0:4250)');

  videoActionHello.parseTimingAction('0:00:30');
  assert.equal(videoActionHello.getTiming(), 30, 'Parse timing (0:00:30)');

  videoActionHello.parseTimingAction('00:00:40');
  assert.equal(videoActionHello.getTiming(), 40, 'Parse timing (00:00:40)');

  videoActionHello.parseTimingAction('00:0:32');
  assert.equal(videoActionHello.getTiming(), 32, 'Parse timing (00:0:32)');

  videoActionHello.parseTimingAction('0:10:00');
  assert.equal(videoActionHello.getTiming(), 600, 'Parse timing (0:10:00)');

  videoActionHello.parseTimingAction('0:10:0');
  assert.equal(videoActionHello.getTiming(), 600, 'Parse timing (0:10:0)');

  videoActionHello.parseTimingAction('00:10:0');
  assert.equal(videoActionHello.getTiming(), 600, 'Parse timing (00:10:0)');

  videoActionHello.parseTimingAction('1:0:0');
  assert.equal(videoActionHello.getTiming(), 3600, 'Parse timing (1:0:0)');

  videoActionHello.parseTimingAction('1:0:10');
  assert.equal(videoActionHello.getTiming(), 3610, 'Parse timing (1:0:10)');

  // invalid parsing
  var isValid = videoActionHello.parseTimingAction('invalid');
  assert.equal(isValid, false, 'Invalid parsing flag');

  // is timed Action
  var videoActionHello2 = new VideoDirectorJS.VideoAction('10s', function(){});
  assert.equal(videoActionHello2.getIsTimedAction(), true, 'getIsTimedAction (true)');

  var videoActionHello3 = new VideoDirectorJS.VideoAction('play', function(){});
  assert.equal(videoActionHello3.getIsTimedAction(), false, 'getIsTimedAction (false)');
});

/**
 * Testing VideoActions exceptions
 */
QUnit.test('Test VideoActions exceptions', function(assert) {
  //
  // Invalid timed action
  assert.throws(
    function(){
      new VideoDirectorJS.VideoAction('invalid action', function(){});
    }, 'Action is invalid (Correct time format?)',
    'Invalid timed action passed'
  );
});


videodirector.js
(function(root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define([], factory);
    } else if (typeof exports === 'object') {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        module.exports = factory();
    } else {
        // Browser globals (root is window)
        root.VideoDirectorJS = factory();
    }
}(this, function() {
    "use strict";

    var VideoDirectorJS = {};

  /**
   * VideoAction class
   *
   * @constructor
   */
  function VideoAction(action, func){
    var noneTimeActions = [
      'play', 'playing', 'pause', 'ended',
      'canplay', 'volumechange','seeked',
      'seeking'
    ];

    // action value
    this.action = action;

    // flag to indicate if this is a timed action
    this.timedAction = false;

    // time action needs to be executed
    this.timingAction = 0;

    // function to be called
    this.func = func;

    // check if we are dealing with a timed action
    if(noneTimeActions.indexOf(action) === -1){
      // parse timing action
      var isValid = this.parseTimingAction(action);

      // if we couldn't parse this action
      if(!isValid){
        throw "Action is invalid (Correct time format?)";
      }else{
        this.timedAction = true;
      }
    }
  }

  /**
   * Get action
   * @return {string} name of action
   */
  VideoAction.prototype.getAction = function(){
    return this.action;
  }

  /**
   * Get if action is timed action
   * @return {boolean} flag indicating if this action is a timed action
   */
  VideoAction.prototype.getIsTimedAction = function(){
    return this.timedAction;
  }

  /**
   * Return timing action
   * @return {integer} timing (seconds) when the action needs to be triggered
   */
  VideoAction.prototype.getTiming = function(){
    return this.timingAction;
  }

  /**
   * Execute callback
   */
  VideoAction.prototype.executeCallback = function(){
    // set executed flag
    this.isExecuted = true;

    // call function
    return this.func();
  }

  /**
   * Parse timing action
   * @param  {string} action timing action
   * @return {boolean} flag to indicate if a time was parsed succesfully
   */
  VideoAction.prototype.parseTimingAction = function(action){
    var isValid = false;
    var m = null;

    //
    // {xx}s
    var re = /(.*)s/;

    if ((m = re.exec(action)) !== null) {
      this.timingAction = parseInt(m[1]);
      isValid = true;
    }

    //
    // {xx}m
    var re = /(.*)m/;

    if ((m = re.exec(action)) !== null) {
      this.timingAction = parseInt(m[1] * 60);
      isValid = true;
    }

    //
    // {xx}m {xx}s
    var re = /(.*)m(?:\s*)(.*)s/;

    if ((m = re.exec(action)) !== null) {
      this.timingAction = parseInt(m[1] * 60) + parseInt(m[2]);
      isValid = true;
    }

    //
    // {xx}:{xx}:{xx}
    var re = /(.*)\:(.*)\:(.*)/;

    if ((m = re.exec(action)) !== null) {
      this.timingAction = parseInt(m[1] * 3600) + parseInt(m[2] * 60) + parseInt(m[3]);
      isValid = true;
    }

    return isValid;
  }

  /**
   * Director class
   *
   * @param {string} videoElemId id of video element
   * @constructor
   */
  function Director(videoElemId){
    // current time
    this.currentTime = 0;
    this.prevTime = 0;

    // holds our actions
    this.videoActions = [];

    // holds our timed video actions
    this.timedVideoActions = [];

    // holds flag that  video has been initialised
    this.hasVideoInitialised = false;

    // get 'pointer' to video element
    this.videoElem = document.getElementById(videoElemId);

    // be sure we got an element
    if(!this.videoElem){
      throw "No element found with this id";
    }

    // be sure we are working with a video tag
    if(this.videoElem.nodeName !== 'VIDEO'){
      throw "Element is not a <video> element";
    }

    // set flag that video has initialised
    this.hasVideoInitialised = true;

    // setup video events
    this.setupEvents();
  }

  /**
   * Setup events
   */
  Director.prototype.setupEvents = function(){
    // check if our video has initialised
    if (this.hasVideoInitialised){
      var self = this;

      /**
       * Executed when a video starts to play
       */
      this.videoElem.onplay = function(){
        self.handleVideoEvent('play');
      }

      /**
       * Executed when a video starts to play after having been paused or stopped for buffering
       */
      this.videoElem.onplaying = function(){
        self.handleVideoEvent('playing');
      }

      /**
       * Executed when a video pauses
       */
      this.videoElem.onpause = function(){
        self.handleVideoEvent('pause');
      }

      /**
       * Executed when a video ended
       */
      this.videoElem.onended = function(){
        self.handleVideoEvent('ended');
      }

      /**
       * Executed when the browser can start playing the movie
       */
      this.videoElem.oncanplay = function(){
        self.handleVideoEvent('canplay');
      }

      /**
       * Executed when the audio volume of a video has changed
       */
      this.videoElem.onvolumechange = function(){
        self.handleVideoEvent('volumechange');
      }

      /**
       * Executed when the user starts moving/skipping to a new position in the video
       */
      this.videoElem.onseeking = function(){
        self.handleVideoEvent('seeking');
      }

      /**
       * Executed when the user is finished moving/skipping to a new position in the video
       */
      this.videoElem.onseeked = function(){
        self.handleVideoEvent('seeked');
      }

      /**
       * Executed when video playing position changed
       */
      this.videoElem.ontimeupdate = function(){
        self.handleOnTime();
      }

    }else{
      throw "Video hasn't initialised";
    }
  }

  /**
   * Handle video eventName
   * @param  {string} eventName name of the event
   */
  Director.prototype.handleVideoEvent = function(eventName){
    // get all video actions of type play
    var filteredVideoActions = this.getVideoActionsOfType(eventName);

    // iterate over video actions
    for(var i = 0, l = filteredVideoActions.length; i<l; i++){
        filteredVideoActions[i].executeCallback();
    }
  }

  /**
   * Handle timing actions
   */
  Director.prototype.handleOnTime = function(){

    // be sure we didn't already process this frame
    if(this.prevTime !== Math.floor(this.videoElem.currentTime)){
      // store current Time and prev time
      this.currentTime = Math.floor(this.videoElem.currentTime);
      this.prevTime = this.currentTime;

      // get actions scheduled at this time
      var filteredVideoActions = this.getTimedVideoActionsOfTime(this.currentTime);

      // iterate over video actions
      for(var i = 0, l = filteredVideoActions.length; i<l; i++){
          filteredVideoActions[i].executeCallback();
      }
    }
  }

  /**
   * Get video actions
   * @return {array} video actions
   */
  Director.prototype.getVideoActions = function(){
    return this.videoActions;
  }

  /**
   * Get Time related VideoActions
   * @return {array} of timed videoactions
   */
  Director.prototype.getTimedVideoActions = function(){
    return this.timedVideoActions;
  }

  /**
   * Get video actions of type
   * @param  {string} actionType name of action type
   * @return {array}  list of videoactions of action type
   */
  Director.prototype.getVideoActionsOfType = function(actionType){
    var tempVideoActions = [];

    // iterate over video actions
    for(var i = 0, l = this.getVideoActions().length; i < l; i++){
        // if current video action is the type we need
        if(this.getVideoActions()[i].getAction() === actionType){
          tempVideoActions.push(this.getVideoActions()[i]);
        };
    }

    return tempVideoActions;
  }

  /**
   * Get timed video actions of time
   * @param  {integer} timeSeconds time of actionType
   * @return {array} list of timed video actions
   */
  Director.prototype.getTimedVideoActionsOfTime = function(timeSeconds){
    var tempVideoActions = [];

    // iterate over video actions
    for(var i = 0, l = this.getTimedVideoActions().length; i < l; i++){
      // if the current time has an action scheduled
      if(this.getTimedVideoActions()[i].timingAction === timeSeconds){
        tempVideoActions.push(this.getTimedVideoActions()[i]);
      }
    }

    return tempVideoActions;
  }

  /**
   * Pause video
   */
  Director.prototype.playVideo = function(){
    this.videoElem.play();
  }

  /**
   * Pause video
   */
  Director.prototype.pauseVideo = function(){
    this.videoElem.pause();
  }

  /**
   * Mute video
   */
  Director.prototype.muteVideo = function(){
    if (this.videoElem.muted){
      this.videoElem.muted = false;
    }else{
      this.videoElem.muted = true;
    }
  }

  /**
   * Set video volume
   * @param  {integer} volume range [0 -100]
   */
  Director.prototype.setVolume = function(volume){
    var volumeVal = 0;

    if(volume > 100){
      volumeVal = 100;
    }else{
      volumeVal = volume / 100;
    }

    this.videoElem.volume = volumeVal;
  }

  /**
   * Get current time
   * @return {integer} current position video (in seconds)
   */
  Director.prototype.getCurrentTime = function(){
    return this.currentTime;
  }


  /**
   * At <action> callback
   * @param  {string}   action that needs to be fullfilled
   * @param  {Function} callback function to call
   */
  Director.prototype.at = function(action, callback){
    // create video action
    var videoAction = new VideoAction(action, callback);

    // when action is a timed action
    if(videoAction.getIsTimedAction()){
      // store in our list of timed actions
      this.timedVideoActions.push(videoAction);
    } else {
      // store in videoactions list
      this.videoActions.push(videoAction);
    }
  }



  VideoDirectorJS.Director = Director;
  VideoDirectorJS.VideoAction = VideoAction;

  return VideoDirectorJS;
}));
